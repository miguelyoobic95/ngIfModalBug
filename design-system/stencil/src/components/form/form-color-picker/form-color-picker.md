---
name: Color Picker
category: Form Components
---

## Attributes
|Name|Type|Default|Description|
|---|---|---|---|
|`color`|string|`#ffffff`|the default color of the color picker|
|`hideLabel`|boolean|false|show or hide the label|

### Default

```yoo-form-color-picker-default.html
<div style="padding:10px;">
    <yoo-form-color-picker></yoo-form-color-picker>
</div>
```

### Color

```yoo-form-color-picker.html
<div style="padding:10px;">
    <yoo-form-color-picker color="#fa2367"></yoo-form-color-picker>
</div>
```

### Hidden Label

```yoo-form-color-picker-hide.html
<div style="padding:10px;">
    <yoo-form-color-picker hide-label="true"></yoo-form-color-picker>
</div>
```

## CSS

|Type|Name|Description|
|---|---|---|
|Large|`large`|large color picker |

### Style Variations

```yoo-form-color-picker-style.html
<div style="padding:10px;">
    <yoo-form-color-picker color="#fa2367"></yoo-form-color-picker>
    <yoo-form-color-picker color="#fa2367" class="large"></yoo-form-color-picker>
    <br/>
    <yoo-form-color-picker hide-label="true" color="#fa2367"></yoo-form-color-picker>
    <yoo-form-color-picker hide-label="true" color="#fa2367" class="large"></yoo-form-color-picker>
</div>
```

## Events

|Attr|Description|
|---|---|---|
|`colorSelected`|When the user change the selected color, send the new one as hexadecimal string|