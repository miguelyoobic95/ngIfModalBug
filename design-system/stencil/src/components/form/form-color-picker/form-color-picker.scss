

$color-picker-size: 1.3rem;

:host {
    .color-picker-container{
        align-items: center;
        padding: 0.2rem;
        width: 6rem;
        input[type=text]{
            -webkit-appearance: none;
            background: transparent;
            font-size: 0.8rem;
            width: 4rem;
            outline: none;
            border-color: transparent;
            text-align: center;
        }
        .color-selector{
            height: 1.3rem;
            width: 1.3rem;
            border-radius: 50%;
            border: 1px solid cssvar(dark-40);
            input[type=color]{
                height: 1rem;
                width: 1rem;
                // min-width: $color-picker-size;
                // min-height: $color-picker-size;
                border: none;
                outline: none;
                -webkit-appearance: none;
                background: transparent;
                opacity: 0;
            }    
            input[type=color]::-webkit-color-swatch-wrapper {
                padding: 0px;
                -webkit-appearance: none;
                background: transparent;
            }
            input[type=color]::-webkit-color-swatch {
                border-radius: 50%;
                border: none;
            }
        }
    }
}

:host(.large){
    .color-picker-container{
        .color-selector{
            height: 1.7rem;
            width: 1.7rem;
            min-width: 1.7rem;
            min-height: 1.7rem;
            input[type=color]{
                height: 1.5rem;
                width: 1.5rem;
            }
        }
    }
}