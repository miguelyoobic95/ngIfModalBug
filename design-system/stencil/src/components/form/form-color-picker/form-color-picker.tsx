import { Component, Prop, State, Event, EventEmitter, Element } from '@stencil/core';

export const DEFAULT_COLOR = '#ffffff';

@Component({
    tag: 'yoo-form-color-picker',
    styleUrl: 'form-color-picker.scss',
    scoped: true
})
export class YooFormColorPickerComponent {

    @Prop() color: string = DEFAULT_COLOR;
    @Prop() hideLabel: boolean = false;

    @Event() colorSelected: EventEmitter<string>;

    @State() currentColor: string;

    @Element() host: HTMLStencilElement;

    colorValidation(newValue: string): void {
        let validation = new RegExp('^#(?:[0-9a-fA-F]{3}){1,2}$'); // Regular Expression to validate hexadecimal color
        if (validation.test(newValue) && newValue.length === 7) {
            this.currentColor = newValue;
        } else {
            this.currentColor = DEFAULT_COLOR;
        }
    }

    onInputChange(ev: Event) {
        this.currentColor = (ev.target as any).value; //Use to re-render !! needed to reset input value
        this.colorValidation((ev.target as any).value);
        this.colorSelected.emit(this.currentColor);
    }

    componentWillLoad() {
        this.colorValidation(this.color);
    }

    render(): JSX.Element {
        let colorSelectorStyle = { background: this.currentColor };
        return (
            <div class="color-picker-container" attr-layout="row">
                <div class="color-selector" style={colorSelectorStyle}>
                    <input type="color" value={this.currentColor} onChange={(event) => this.onInputChange(event)} onInput={(event) => this.onInputChange(event)} />
                </div>
                {this.hideLabel ? null :
                    <input type="text" value={this.currentColor} onChange={(event) => this.onInputChange(event)} />
                }
            </div>
        );
    }
}
