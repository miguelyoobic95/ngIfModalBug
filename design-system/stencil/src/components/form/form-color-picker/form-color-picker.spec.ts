import { TestWindow } from '@stencil/core/testing';
import { YooFormColorPickerComponent, DEFAULT_COLOR } from './form-color-picker';

describe('YooFormColorPickerComponent', () => {
    it('should build', () => {
        expect(new YooFormColorPickerComponent()).toBeTruthy();
    });

    describe('Rendering', () => {
        it ('Should render with no prop', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooFormColorPickerComponent],
                html: '<yoo-form-color-picker></yoo-form-color-picker>'
              });
            expect(element).toMatchSnapshot();
        });

        it ('Should render with a color as prop', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooFormColorPickerComponent],
                html: '<yoo-form-color-picker color="#fa2367"></yoo-form-color-picker>'
              });
            expect(element).toMatchSnapshot();
        });

        it ('Should render without label', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooFormColorPickerComponent],
                html: '<yoo-form-color-picker hide-label="true"></yoo-form-color-picker>'
              });
            expect(element).toMatchSnapshot();
        });
    });

    describe('Prop rendering', () => {
        it ('Should re-render when color prop is updated', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooFormColorPickerComponent],
                html: '<yoo-form-color-picker color="#fa2367"></yoo-form-color-picker>'
              });
            expect(element.color).toEqual('#fa2367');
            element.color = '#aa34aa';
            await window.flush();
            expect(element.color).toEqual('#aa34aa');
            expect(element).toMatchSnapshot();
        });

        it ('Should re-render when hide-label prop is updated', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooFormColorPickerComponent],
                html: '<yoo-form-color-picker></yoo-form-color-picker>'
              });
            expect(element.hideLabel).toEqual(false);
            element.hideLabel = true;
            await window.flush();
            expect(element.hideLabel).toEqual(true);
            expect(element).toMatchSnapshot();
        });
    });

    describe('Prop Validation', () => {
        it('Should use default color if the color is not a valid color', async () => {
            let element = new YooFormColorPickerComponent();
            element.colorValidation('someColor');
            expect(element.currentColor).toEqual(DEFAULT_COLOR);
            element.colorValidation('wrongColor');
            expect(element.currentColor).toEqual(DEFAULT_COLOR);
        });

        it('Should change color if input is a valid color', async () => {
            let element = new YooFormColorPickerComponent();
            element.colorValidation('#aacc23');
            expect(element.currentColor).toEqual('#aacc23');
        });
    });

   describe('events', () => {
       let element = new YooFormColorPickerComponent();
       element.colorSelected = {
           emit: () => {}
       };

       it('should emit event on input change', () => {
        let res = '#aacc23';
        const colorSelectedSpy = jest.spyOn(element.colorSelected, 'emit');
        element.onInputChange({target: {value: '#aacc23'}});
        expect(element.currentColor).toEqual(res);
        expect(colorSelectedSpy).toHaveBeenCalledWith(res);
       });
   });
});
