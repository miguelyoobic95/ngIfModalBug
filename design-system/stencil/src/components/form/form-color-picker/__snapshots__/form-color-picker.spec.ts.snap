// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YooFormColorPickerComponent Prop rendering Should re-render when color prop is updated 1`] = `
<yoo-form-color-picker
  class="hydrated"
  color="#fa2367"
  data-yoo-form-color-picker-host=""
>
  
  <div
    attr-layout="row"
    class="color-picker-container"
    data-yoo-form-color-picker=""
  >
    <div
      class="color-selector"
      data-yoo-form-color-picker=""
      style="background: rgb(250, 35, 103);"
    >
      <input
        data-yoo-form-color-picker=""
        type="color"
      />
    </div>
    <input
      data-yoo-form-color-picker=""
      type="text"
    />
  </div>
</yoo-form-color-picker>
`;

exports[`YooFormColorPickerComponent Prop rendering Should re-render when hide-label prop is updated 1`] = `
<yoo-form-color-picker
  class="hydrated"
  data-yoo-form-color-picker-host=""
>
  
  <div
    attr-layout="row"
    class="color-picker-container"
    data-yoo-form-color-picker=""
  >
    <div
      class="color-selector"
      data-yoo-form-color-picker=""
      style="background: rgb(255, 255, 255);"
    >
      <input
        data-yoo-form-color-picker=""
        type="color"
      />
    </div>
    
  </div>
</yoo-form-color-picker>
`;

exports[`YooFormColorPickerComponent Rendering Should render with a color as prop 1`] = `
<yoo-form-color-picker
  class="hydrated"
  color="#fa2367"
  data-yoo-form-color-picker-host=""
>
  
  <div
    attr-layout="row"
    class="color-picker-container"
    data-yoo-form-color-picker=""
  >
    <div
      class="color-selector"
      data-yoo-form-color-picker=""
      style="background: rgb(250, 35, 103);"
    >
      <input
        data-yoo-form-color-picker=""
        type="color"
      />
    </div>
    <input
      data-yoo-form-color-picker=""
      type="text"
    />
  </div>
</yoo-form-color-picker>
`;

exports[`YooFormColorPickerComponent Rendering Should render with no prop 1`] = `
<yoo-form-color-picker
  class="hydrated"
  data-yoo-form-color-picker-host=""
>
  
  <div
    attr-layout="row"
    class="color-picker-container"
    data-yoo-form-color-picker=""
  >
    <div
      class="color-selector"
      data-yoo-form-color-picker=""
      style="background: rgb(255, 255, 255);"
    >
      <input
        data-yoo-form-color-picker=""
        type="color"
      />
    </div>
    <input
      data-yoo-form-color-picker=""
      type="text"
    />
  </div>
</yoo-form-color-picker>
`;

exports[`YooFormColorPickerComponent Rendering Should render without label 1`] = `
<yoo-form-color-picker
  class="hydrated"
  data-yoo-form-color-picker-host=""
  hide-label="true"
>
  
  <div
    attr-layout="row"
    class="color-picker-container"
    data-yoo-form-color-picker=""
  >
    <div
      class="color-selector"
      data-yoo-form-color-picker=""
      style="background: rgb(255, 255, 255);"
    >
      <input
        data-yoo-form-color-picker=""
        type="color"
      />
    </div>
    
  </div>
</yoo-form-color-picker>
`;
